src/Compilers/CSharp/Portable/Symbols/TupleTypeSymbol.cs

```diff
@@ -73,31 +82,154 @@ internal static TupleTypeSymbol ConstructTupleTypeSymbol(TupleTypeSymbol origina
+            var chainBuilder = ArrayBuilder<MethodSymbol>.GetInstance(nestings);
 
-            for (int i = 0; i < numElements; i++)
+            for (int i = 0; i < nestings; i++)
             {
-                var wellKnownTypeMember = GetTupleTypeMember(numElements, i);
-                var underlyingField = (FieldSymbol)Binder.GetWellKnownTypeMember(binder.Compilation, wellKnownTypeMember, diagnostics, syntax: syntax);
-                var field = new TupleFieldSymbol(elementNames.IsDefault ? UnderlyingMemberName(i) : elementNames[i],
-                                           this,
-                                           elementTypes[i],
-                                           underlyingField?.AsMember(underlyingType));
-                fieldSymbols.Add(field);
+                var ctor = (MethodSymbol)Binder.GetWellKnownTypeMember(binder.Compilation, TupleTypeSymbol.GetTupleCtor(nextArity), diagnostics, syntax: node);
+                if ((object)ctor != null)
+                {
+                    ctor = ctor.AsMember(UnderlyingTupleTypeChain[UnderlyingTupleTypeChain.Count - i - 1]);
+                    chainBuilder.Add(ctor);
+                }
+                else
+                {
+                    hasErrors = true;
+                }
+                nextArity = 8;
             }
 
-            fields = fieldSymbols.ToImmutableAndFree();
+            chainBuilder.ReverseContents();
+            return chainBuilder.ToImmutableAndFree();
+        }
 
-            return underlyingType;
+        /// <summary>
+        /// Returns the number of chunks required to represent numElements as nested ValueTuples.
+        /// For example, for 8 elements, you need 2 ValueTuples.
+        /// </summary>
+        static internal int NumberOfValueTuples(int numElements)
+        {
+            return (numElements - 1) / 7 + 1;
+        }
+        /// <summary>
+        /// Returns the size of the last ValueTuple.
+        /// For example, for 9 elements, the last chunk's size is 2.
+        /// </summary>
+        static internal int SizeOfLastValueTuples(int numElements)
+        {
+            return (numElements - 1) % 7 + 1;
+        }
+
+        /// <summary>
+        /// Produces the underlying ValueTuple corresponding to this list of element types.
+        ///
+        /// For 2 elementTypes of type Int32, returns ValueTuple&lt;Int32, Int32>
+        /// For 9 elementTypes, returns ValueTuple&lt;Int32, ..., ValueTuple&lt;Int32, Int32>>, and ValueTuple&lt;Int32, Int32> (largest first)
+        /// </summary>
+        private static OneOrMany<NamedTypeSymbol> GetTupleUnderlyingTypeChain(ImmutableArray<TypeSymbol> elementTypes, CSharpSyntaxNode syntax, Binder binder, DiagnosticBag diagnostics)
+        {
+            var numElements = elementTypes.Length;
+            var numChunks = NumberOfValueTuples(numElements);
+            var nextChunkSize = SizeOfLastValueTuples(numElements);
+            if (numChunks == 1)
```

Comment:
It feels like once we get rid of requirement to build the chain as a separate data structure, the code in this function could be significantly simplified and can use one loop.